### ISSUE TITLE
Reproducible content

### ISSUE THREAD

**Issue by glerchundi** [bitnami-labs/sealed-secrets#131]
Is it possible to make generation reproducible so that the generated code is the same between runs if the origin if the same?

**Comment by glerchundi** [comment]
Umm, as far as i can see there is a random seal-time key generation [here](https://github.com/bitnami-labs/sealed-secrets/blob/master/pkg/crypto/crypto.go#L26). There are two possibilities to achieve this:
1. Change current crypto algorithm, which would break compatibility with old `SealedSecret` resources. It's alpha so it's expected ;)
2. Add a hash/checksum whatever info to the `SealedSecret` and block sealing unless the content is different or a `--force` like argument is provided

**Comment by anguslees** [comment]
This would be a severe security risk, since it makes it possible to
- Notice when a secret is set to a known plaintext value, by just comparing the ciphertext
- Build tables of known values and do rainbow attacks

Since the algorithm and public key are public, the random salt is important to make it hard to do known-plaintext attacks.

**Comment by glerchundi** [comment]
@anguslees i'm going to explain my workflow right now to show you how i'm achieving this:

Here it is the Kubernetes Secret template:

```
{
    "apiVersion": "v1",
    "kind": "Secret",
    "metadata": {
        "name": "registry-gcr-io"
    },
    "type": "kubernetes.io/dockerconfigjson",
    "data": {
        ".dockerconfigjson": "{{encodeBase64 (sopsDecrypt ".dockerconfigjson.enc")}}"
    }
}
```

I'm using a tool I built to make this workflow work which is no more than a template renderer with handy functions on it, one of those is the `sopsDecrypt` which aids me grouping the different secrets into one Kubernetes Secret, take a look here: https://github.com/glerchundi/tmpl-renderer.

I slightly modified the script i'm currently using to support this 'reproducible sealing':

```
#!/bin/bash

set -euo pipefail

ext=".secret.json.tmpl"
for src in $(find . -name *${ext}); do
    dst=$(dirname ${src})/$(basename ${src} ${ext}).sealedsecret.json
    dstsum=${dst}.sum

    plaintext=$(tmpl-renderer ${src} | jq .)
    sum=$(echo -n "${plaintext}" | md5)
    if [ -f ${dstsum} ] && [ "${sum}" = "$(cat ${dstsum})" ]; then
        echo "Ignoring ${src} as the content has not changed."
        continue
    fi

    echo "Generating '${dst}' from '${src}'..."
    tmpl-renderer ${src} | jq . | kubeseal --controller-name sealed-secrets | jq . > ${dst}
    echo -n "${sum}" > ${dstsum}
done
```

Which is no more than: "Generate if and only if the original secret has changed". In case you see any weakness in the whole process it would be really helpful to know that!

Hope it helps.

**Comment by anguslees** [comment]
Yep, your script looks fine to me.  Doing a content-check for _when_ to regenerate the SealedSecret is a reasonable thing to do.   I think the specifics of how/when to do that are going to vary based on the desired change workflow, so are best managed in the surrounding build tool (like make, or one of the newer modern alternatives) rather than within `kubeseal` itself.

If you have a suggested wrapper shell script that you're happy with, then we could add it to the sealed-secrets repo as a `contrib` script if you wish.
